File: OrchidCore/src/main/java/com/eden/orchid/impl/themes/functions/FindAllFunction.java
package com.eden.orchid.impl.themes.functions;

import com.eden.orchid.api.OrchidContext;
import com.eden.orchid.api.compilers.TemplateFunction;
import com.eden.orchid.api.indexing.IndexService;
import com.eden.orchid.api.options.annotations.Description;
import com.eden.orchid.api.options.annotations.IntDefault;
import com.eden.orchid.api.options.annotations.Option;
import lombok.Getter;
import lombok.Setter;

import javax.inject.Inject;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

@Getter @Setter
@Description(value = "Get all Page objects matching a query.", name = "Find all")
public final class FindAllFunction extends TemplateFunction {

    private final OrchidContext context;

    @Option
    @Description("The Id of the items to link to.")
    private String itemId;

    @Option
    @Description("The type of collection the items are expected to come from.")
    private String collectionType;

    @Option
    @Description("The specific Id of the given collection type where the items are expected to come from.")
    private String collectionId;

    @Option("page") @IntDefault(0)
    @Description("Paginate results starting at this page.")
    public int pageIndex;

    @Option @IntDefault(0)
    @Description("Paginate results using this as a page size")
    private int pageSize;

    @Inject
    public FindAllFunction(OrchidContext context) {
        super("findAll", false);
        this.context = context;
    }

    @Override
    public String[] parameters() {
        List<String> params = new ArrayList<>(Arrays.asList(IndexService.locateParams));
        params.add("page");
        params.add("pageSize");
        return params.toArray(new String[0]);
    }

    @Override
    public Object apply() {
        if(pageIndex > 0 && pageSize > 0) {
            return context.findAll(collectionType, collectionId, itemId, pageIndex, pageSize);
        }
        else {
            return context.findAll(collectionType, collectionId, itemId);
        }
    }

}

File: OrchidCore/src/main/kotlin/com/eden/orchid/impl/themes/functions/FindAllFunction.kt
package com.eden.orchid.impl.themes.functions

import com.eden.orchid.api.OrchidContext
import com.eden.orchid.api.compilers.TemplateFunction
import com.eden.orchid.api.indexing.IndexService
import com.eden.orchid.api.options.annotations.Description
import com.eden.orchid.api.options.annotations.IntDefault
import com.eden.orchid.api.options.annotations.Option
import com.eden.orchid.api.theme.pages.OrchidPage

@Description(value = "Get all Page objects matching a query.", name = "Find all")
class FindAllFunction : TemplateFunction("findAll", false) {

    @Option
    @Description("The Id of the items to link to.")
    lateinit var itemId: String

    @Option
    @Description("The type of collection the items are expected to come from.")
    lateinit var collectionType: String

    @Option
    @Description("The specific Id of the given collection type where the items are expected to come from.")
    lateinit var collectionId: String

    @Option
    @IntDefault(0)
    @Description("Paginate results starting at this page. Index starts at 1.")
    var pageIndex: Int = 0

    @Option
    @IntDefault(0)
    @Description("Paginate results using this as a page size")
    var pageSize: Int = 0

    override fun parameters(): Array<String> {
        return arrayOf(
                *IndexService.locateParams,
                "page",
                "pageSize",
                "pageIndex"
        )
    }

    override fun apply(context: OrchidContext, page: OrchidPage?): Any? {
        return if (pageIndex > 0 && pageSize > 0) {
            context.findAll(collectionType, collectionId, itemId, pageIndex, pageSize)
        } else {
            context.findAll(collectionType, collectionId, itemId)
        }
    }
}
